### DISCUSSION TITLE
Approaches for template generation integrated tests and CI?

### DISCUSSION THREAD

**Discussion by manoelpqueiroz** [cookiecutter/cookiecutter#2109] [Q&A]
Hi guys!

Has anyone worked with integrated tests for their templates, validating that a project generated from a Cookiecutter template is satisfactorily rendering as expected?

I'm interested especially in running this kind of testing in CI/CD pipelines for my [template](https://gitlab.com/manoelpqueiroz/galactipy). Currently the CI deals with testing and coverage of functions defined in `hooks` modules. My idea is that I can extend it so, for instance, I can not only run the template's current `tests`, but also ensure my desired output for the generated project's `tests` as well.

Basically, I am trying to find a way to use a single CI pipeline to generate a project and execute other CI tasks especifically for the generated project. Otherwise, I am restricted to developing my template and setting up additional repositories to mirror possible generated projects and test their integrity individually.

I am aware of the `pytest-cookies` package for testing correct outputs of a template, but I have never been able to make it work with my case and it seems the project has been abandoned, which doesn't bode well for security.

Has anyone here faced a similar challenge? I would love to hear your ideas on this topic.

**Comment by mikeweltevrede** [comment]
You can create a fixture to use the `cookiecutter` command to generate the project (I guess `pytest-cookies` might do the same, didn't take a look). Your test in the root repo would then run `pytest` with the `pytest.main` function. The disadvantage of this, however, is that it cannot run the tests of the project to be generated individually and you would only have one potentially long running test to run all the project's tests.
